### DISCUSSION TITLE
What's the different of \CallBackFunctionSample and \CallBackWindowSample demo?

### DISCUSSION THREAD

**Discussion by haohuancheng** [FactoryXCode/MfPack#23] [Q&A] [answered]
and , how to replay the capture  MIC audio to speaker but not discards them ? thank you!

**Comment by FactoryXCode** [comment]
The CallBackWindowSample uses the CALLBACK_WINDOW callbacktype as used in the original C++ sample. This means it will handle all messages from a window using its handle (HWND). The latter uses a callback procedure address to open a device.
You could -for example- render the samples to a file (wav) and replay, if that is what you mean. Microsoft created some samples about how to do that. For example the [Audio Streaming Sample Code](https://docs.microsoft.com/en-us/windows/win32/directshow/audio-streaming-sample-code)  or using this wasapi sample: [Capturing a Stream](https://docs.microsoft.com/en-us/windows/win32/coreaudio/capturing-a-stream).

**Comment by haohuancheng** [comment]
But, in these two demo program(CallBackFunctionSample and CallBackWindowSample ),   How can I experience the effect of chat. speak to mic and replay to speaker?  Is there some code missing? I want to use this demonstration program to write a network real-time voice chat program. Can it be carried out on the basis of these codes? What more work needs to be done? Now, I don't know what demonstration these two programs do.

**Comment by FactoryXCode** [comment, marked as answer]
In the code you should read the comments. The sample is about how to use the ducking features using wav or wasapi, that's up to you, not how to implement the features you want. Writing a complete code on demand is not our goal, sorry, we don't have time for that.
